build_runner/lib/src/entrypoint/test.dart

```diff
@@ -136,8 +146,30 @@ class TestCommand extends BuildRunnerCommand {
           '--precompiled',
           precompiledPath,
         ]..addAll(extraTestArgs),
-        mode: ProcessStartMode.inheritStdio);
-    _ensureProcessExit(testProcess);
+        mode: shouldCaptureOutput
+            ? ProcessStartMode.normal
+            : ProcessStartMode.inheritStdio);
+    _ensureProcessExit(testProcess, forwardExitSignals: shouldCaptureOutput);
+    if (shouldCaptureOutput) {
+      try {
+        final futures = [
+          // Forward stderr so that non-test errors are still visible.
+          stderr.addStream(testProcess.stderr),
+        ];
+        if (testStdout != null) {
+          var sink = testStdout.openWrite();
+          futures.add(sink
+              .addStream(testProcess.stdout)
+              .then((_) => sink.flush)
+              .then((_) => sink.close));
+        }
+        await Future.wait(futures);
+        stdout.writeln('Test results written to ${testStdout.path}');
+      } catch (e) {
```

Comment:
What types of exceptions might we catch here?

Reply:
Probably just `FileSystemException`s, potentially when creating the file. I ended up having to update line 160 locally to support writing to a directory that doesn't yet exist:

```
var sink = (testStdout..createSync(recursive: true)).openWrite();
```